### PR TITLE
Dirent lookup speedup

### PR DESCRIPTION
**Pull request by veloman-yunkan**
This PR addresses #410 & #424

The speed-up of dirent look-up is achieved via an auxiliary special purpose in-memory cache (new helper class `zim::NarrowDown`). It stores the full urls (ns+path) and indices of uniformly spaced dirents and is filled upon opening the ZIM file. Then the dirent look-up operation is performed in two stages:

1. Narrow down the range that might contain the dirent with the given full url via a binary search in that cache.
2. Perform binary search on dirents (with disc access as needed) in that narrower range.

Since the speed-up is due to avoiding expensive IO operations, the goal is to fit as much data as possible in the in-memory cache. This is achieved as follows:
1. A sorted vector container is used instead of a tree (`std::map`) data structure.
2. Actual URLs (ns+path) are approximated with shortened versions that allow to narrow-down the initial full range with almost the same efficiency except a few corner cases (see below for the explanation of this trick).
3. (The shortened) URLs are not stored as `std::string` objects (which have an overhead of a couple of pointers). Instead they are packed sequentially as C-style strings in a single array, and 4-byte offsets into that array are used.

The idea of shortening the URLs is illustrated on an example:

| Dirent #  |  Full URL (namespace concatenated with path)      |
|-------------|--------------------------------------------------------------------|
| ...           |         ...                                                                      |
| 1234      |  AList_of_Climate_Change_Initiatives                      |
| 1235      |  AList_of_Greta_Thunberg_speeches                      |
| ...           |         ...                                                                      |

The example above illustrates two consecutive entries from the wikipedia_en_climate_change_nopic_2020-01.zim ZIM file (though the index values are not real). If we were to use the item `#1234` in our cache the naive approach would be to store its full URL as is. However, let's imagine that the ZIM file also contained an item with full URL **AList_of_G**. Then it would have to reside between **AList_of_Climate_Change_Initiatives** and **AList_of_Greta_Thunberg_speeches**. So we can pretend that such an item exists and store in our cache the fictitious entry (**AList_of_G**, 1234.5). When we arrive at that entry during the range narrow-down step we must round the dirent index downward if it is going to be used as the lower bound of the range, and round it upward if it is going to be used as the upper bound of the range.

---

## Benchmark results

Since this optimization effectively adds 1024 additional (but static) dirent lookup entries, the benchmark includes two runs of the `master` version with different values for the `ZIM_DIRENTCACHE` setting - 512 (the default) and 1536 (default + 1024).

In summary, this optimization improves the performance of `zimcheck` more than simply increasing the dirent cache size does, and doesn't negatively impact the performance of `kiwix-serve`.

Data used for benchmarking:

| id |  ZIM file                                                                     |    size (MB) | article count | cluster count |
|----|---------------------------------------------------------------------|----------------|------------------|-------------------|
| 1 | wikipedia_en_climate_change_nopic_2020-01.zim  |  31    | 7646     | 51     |
| 2 | [wikipedia_hy_all_mini_2020-07.zim][2]                    |   560 | 509325 | 1518 |
| 3 | [wikipedia_hy_all_mini_2020-08.zim][1]                    |   563 | 509611 | 1526 |

---

### `zimcheck -A` benchmark

|   ZIM file id  | zimcheck -A runtime (master, DIRENTCACHE=512) | zimcheck -A runtime (master, DIRENTCACHE=1536) |  zimcheck -A runtime (this PR, DIRENTCACHE=512) | zimcheck -A runtime (this PR, DIRENTCACHE=1536) |
|-----|-----------------|-----------------|---------------|-----------------|
|  1  | 6.183s        | 5.881s        | 5.693s        |  5.531s       |
|  2  | 6m36.615s | 6m12.248s | 5m51.896s | 5m40.461s |
|  3  | 6m31.202s | 6m16.159s | 5m55.336s | 5m42.636s |

---
### `kiwix-serve` benchmark

The benchmark was performed using the [`benchmark_kiwix_serve`](https://github.com/veloman-yunkan/kiwix-dev-utils/blob/80562e5de15faba59dfec1c3a2da147a699825c9/benchmark_kiwix_serve) script as follows:

```
ZIM_DIRENTCACHE=512 ./benchmark_kiwix_serve master 1000 zimfiles/wikipedia_hy_all_mini_2020-08.zim
ZIM_DIRENTCACHE=1536 ./benchmark_kiwix_serve master 1000 zimfiles/wikipedia_hy_all_mini_2020-08.zim
ZIM_DIRENTCACHE=512 ./benchmark_kiwix_serve pr430 1000 zimfiles/wikipedia_hy_all_mini_2020-08.zim
```

That script starts a kiwix server and fetches from it about 1000 articles using `wget --recursive -l 1`.

|   ZIM file id     | wget runtime (master, DIRENTCACHE=512) | wget runtime (master, DIRENTCACHE=1536) |  wget runtime (this PR, DIRENTCACHE=512) |  wget runtime (this PR, DIRENTCACHE=1536) |
|--------------------|------------|-------------|------------|-------------|
|  3                    | 36.621s | 35.628s | 36.210s | 36.236s |

(It was verified that the output directories created by wget from both runs are identical, with the exception of `/random`)

[1]: http://download.kiwix.org/zim/wikipedia/wikipedia_hy_all_mini_2020-08.zim
[2]: http://download.kiwix.org/zim/wikipedia/wikipedia_hy_all_mini_2020-07.zim

### CODE REVIEW

**Comment by mgautierfr** [comment]
Before you go too further in this PR, it would be nice to do two PR. One moving the existing code into the new DirentLookup (without change in the algorithm). And another one adding the "fanout" algorithm (and it seems you don't follow the description provided in #424. Whatever is the best idea, we will probably discuss here).

**Comment by veloman-yunkan** [comment]
@kelson42 @mgautierfr 

Please see the benchmarking results of this draft PR before deciding how to proceed from here.

**Comment by kelson42** [comment]
@veloman-yunkan Thank you for the benchmark. If I understand properly, from a speed perspective we should better have a big dirent cache, but what does that imply in term of memory consumption? How much additional memory is needed to store 1024 Dirent entries?

**Comment by mgautierfr** [comment]
> Please see the benchmarking results of this draft PR before deciding how to proceed from here.

The idea of splitting the PR is still valid. The first PR should not change the performance (but improve the code organization) and the second PR will work on performance (and let us discuss about different algorithm possibilities)

---

I'm a bit surprised by the benchmark results.
zimcheck loop over the dirent without using the `find` method (binary search) and this PR (the performance part) is reducing the range in which search. So the performance should be on the kiwix-serve side, not the zimcheck.

**Comment by veloman-yunkan** [comment]
> I'm a bit surprised by the benchmark results.
> zimcheck loop over the dirent without using the `find` method (binary search)

`zimcheck -A` includes URL validation which [checks that all internal links in the ZIM file are valid](https://github.com/openzim/zim-tools/blob/a46e5d935001e6c2ed90fd2094d6d1355621c912/src/zimcheck/checks.cpp#L182-L187).

**Comment by veloman-yunkan** [comment]
> To be complete, it would be really helpful to have your PR with 1536 Dirents in the cache.

Done

**Comment by kelson42** [comment]
@mgautierfr @veloman-yunkan Thank you. I'm still unsure about the memory impact of `DIRENTCACHE=1536`, but if this is not taking too much memory, we should better use it?

**Comment by veloman-yunkan** [comment]
> If I understand properly, from a speed perspective we should better have a big dirent cache, but what does that imply in term of memory consumption? How much additional memory is needed to store 1024 Dirent entries?

@kelson This PR doesn't cache the dirents. It caches only data needed for the initial stage of the lookup. In the current prototype version the item URL/path is fully cached. This proof-of-concept is not optimized for memory usage, and there is a way to reduce it significantly. A finely tuned implementation will use a a couple of dozen bytes per lookup cache entry.

**Comment by veloman-yunkan** [comment]
> I'm still unsure about the memory impact of `DIRENTCACHE=1536`, but if this is not taking too much memory, we should better use it?

1536 is not a magic number and hardly it is the sweet spot for a typical runtime environment. Using a larger number will likely result in even more speed-up on average.

**Comment by kelson42** [comment]
> > I'm still unsure about the memory impact of `DIRENTCACHE=1536`, but if this is not taking too much memory, we should better use it?
> 
> 1536 is not a magic number and hardly it is the sweet spot for a typical runtime environment. Using a larger number will likely result in even more speed-up on average.

I had a bit the feeling the speed progression here might be asymptotic (and not linear) and therefore we could keep a pre-defined value which works pretty well (but not optimaly) everywhere. Otherwise we would be back to our discussion on #311 I believe. I will let you sort out the details with @mgautierfr. Just wanted to emphasis that (1) the improvements seem good (2) if the memory cost is not too big, we should not be afraid to increase significantly the `DIRENTCACHE` size.

**Comment by veloman-yunkan** [comment]
@kelson42 This PR will greatly reduce the utility from increasing the size of the general purpose dirent cache. Since dirent cache was heavily used during dirent look-up operations, we better have a dedicated, more optimal cache serving that operation better.

**Comment by mgautierfr** [comment]
In this PR you use a different approach of what I expose in #424.

Your implementation a 1024 items map and split the whole articles range into 1024 balanced subranges.
But the lookup for the subrange imply a `upper_bound` call (`O(log(1024))`, so `O(1)` as 1024 is const) which is a bit longer than a direct array indexing.

My proposition split the range in 256 sub-range (not balanced) (but allow subdividing sub-range into 256 sub-sub-range). It is a bit more complex to construct (as we build the subranges based on the content and so we have to parse the content). But the lookup is probably a bit faster as we do a simple array indexing. It also has the advantage to store less data (a 4 bytes offset per range) compared to yours (a 4 bytes offset and the dirent path).

It seems that your implementation is easier to implement (it is already implemented) and build the cache array faster. But it is potentially slower to lookup dirents once the cache is created. It seems my idea is better for long running process. (But yet again, I don't what is the real performance gain between the two ideas)

As you already said, your PR is draft and not optimized for memory usage. But I would like to know your point here. Does your algorithm is the "final" idea or it is just a first implementation of a array cache to have benchmark numbers and them we see what is best to implement ?

**Comment by veloman-yunkan** [comment]
@mgautierfr 

> Does your algorithm is the "final" idea or it is just a first implementation of a array cache to have benchmark numbers and them we see what is best to implement ?

This is only a prototype. I definitely know how to improve the memory consumption per cache item, so that we can afford more of them. But that preserves one drawback of this implementation - slow startup due to eagerly reading all those dirents. How serious is that issue? Since kiwix-serve can be started with a library consisting of a lot of ZIM files, is it acceptable to initialize such a lookup data structure eagerly? Should we instead try to fill it on demand?

**Comment by mgautierfr** [comment]
>  How serious is that issue? 

I don't know :) 
Your current implementation is probably not a issue. The cache filling is just 1024 dirent reads.
While it takes time, it is not so much. (The actual implementation searching for start of namespaces with binary search may reads dozens of dirents, and my algo will probably need hundreds)

> Since kiwix-serve can be started with a library consisting of a lot of ZIM files, is it acceptable to initialize such a lookup data structure eagerly? Should we instead try to fill it on demand?

On kiwix-serve, I would say it is not a issue as we are in a long running process. The warmup time will not be so important, we will simply have to wait a bit longer before the server is up.
On a one shoot process, it may be more important, but it would be on only one zim.

On kiwix-serve, we open all zim files only to locate the xapian database to have multizim search. Else we would open zim file only on demand. I think we should have different strategy depending of the usage.
(Opening zim file without cache to locate xapian database or zimdump, Opening with cache for long process, ...). But this strategy selection will not be made here. For now, I think we can assume that the warmup time is not a real issue (if it is reasonable).

**Comment by kelson42** [comment]
@veloman-yunkan Hi, any news here? We wait to that PR to:
* complete speedup project on libzim
* make a new release of lizim and zimtools to benefit of the full speedup

**Comment by kelson42** [comment]
@veloman-yunkan Any timeline for PR completion?

**Comment by veloman-yunkan** [comment]
@kelson42 I will try to complete it next week.

File: src/narrowdown.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    if ( it != entries.begin() )
+      --it;
+    r.begin = it->upperBound;
+    return r;
+  }
+
+private: // types
+  struct Entry
+  {
+    std::string keyPrefix;
```
Comment (mgautierfr) [inline src/narrowdown.h:64]: This is not a key prefix. This is "simply" a key. We are not looking a the value start with the prefix but simply compare the value with the key.

File: src/narrowdown.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    // The index of the earliest item with key larger than keyPrefix
+    article_index_type upperBound;
```
Comment (mgautierfr) [inline src/narrowdown.h:67]: From this comment (and `NarrowDown` code), the `upperBound` should be the index *after* the key.

But in `DirentLookup` you fill the lookupGrip with `lookupGrid.add(getDirentKey(i), i);`.
Shouldn't be `lookupGrid.add(getDirentKey(i), i+1)` to set the `upperBound` after the key ?

File: src/dirent_lookup.h [hunk 3]
```diff
@@ -91,93 +147,39 @@ DirentLookup<Impl>::getNamespaceRangeEnd(char ns) const
...
     article_index_type p = l + (u - l) / 2;
-    auto d = impl.getDirent(article_index_t(p));
+    const auto d = impl->getDirent(article_index_t(p));
 
-    int c = ns < d->getNamespace() ? -1
-          : ns > d->getNamespace() ? 1
-          : url.compare(d->getUrl());
+    const int c = ns < d->getNamespace() ? -1
+                : ns > d->getNamespace() ? 1
+                : url.compare(d->getUrl());
```
Comment (mgautierfr) [inline src/dirent_lookup.h:168]: Here is something I potentially wanted to change but never does :
As the range (l-r) is (was) namespace boundary it should be useless to check the namespace. We could simply compare the url.

This PR remove the fact that the range contains only one namespace so we still need to keep the namespace compare. 
Maybe we can go with a more common idea and keep track of the prefix length we are sure to have (if `l` and `r`'s urls start with the same N char, we know that the middle dirent will do the same, we can start the url comparison from array N)
But I'm not sure that keeping track of the prefix is more efficient than a simple comparison.

**Review by mgautierfr — COMMENTED** [review]
Here few questions/comments. Some of them are really about your code. Some others are more general though I think about while reading this PR, we may discuss about them but they don't need a answer now.

---

On a more global way ,it worth to mention this PR do not change the global algorithm, it is more like a kind of improved/efficient/specialized cache system :
- At init we are reading `DIRENT_LOOKUP_CACHE_SIZE` (default 1024) dirents and put the "ns+url" inside a vector.
- At reading we are doing binary search in the vector to find a range and then continue the binary search inside this range. The first search in the vector is fast as the data is already preloaded (this is the cache). The second search use the same code as before (slow as we have to load the pivot dirent (still using the "generic" dirent cache).

So this PR improve the search by using a efficient cache (probably the best we can have without changing the whole algorithm), but we are still doing a binary search on the whole dirents.

This point is not a problem and it will not impact the "mergeability" of this pr but I would like to mention it anyway :)

And it opens to another questions ?
- Should we remove (or at least rework) the generic dirent cache system as we have this efficient one ?
- Should we do the same for title search ?

**Comment by Jaifroid** [comment]
In Kiwix JS, we are about to merge a DirEntry block cache originally developed by peter-x (https://github.com/kiwix/kiwix-js/pull/658), which uses a Least Recently Used algorithm (for discarding cached clusters). We have set the caceh at 4000 blocks of 4096 bytes each. It massively speeds up binary search. A near-case-sensitive title pointer list search for "the second world war experience" in full English Wikipedia, searching for three sensible combinations per word (e.g. second, Second, SECOND) requires 3^5 = 243 searches for different combinations, generating a total of >10,000 cache hits (just for the search). 

Without the block cache, such a search can take 15 seconds. With the block cache, after the very first combination search, the cache already has the first few hops remembered, and the total search of >10,000 lookups takes ~2s. For the last 6,000 lookups, the cache hit rate is 100%. For the first 1,500 the hit rate is 74%, and for the next 2,000 it is 96.8%. The chosen article then loads almost instantaneously.

Remember this is all in pure JavaScript, which you would expect to be slow (binary search is **not** part of the Emscripten port of C++ code from `xzdec` or `zstddec`).

All of this to say that some simple caching strategies, whcih effectively cache the initial hops automatically without the need for pre-computing hops in mwoffliner, can yield very good results.

File: src/narrowdown.h (hunk 1 above)
Comment (veloman-yunkan) [reply src/narrowdown.h]: I named the variable for the future. In later commits shortened pseudo-keys are stored in `zim::NarrowDown`. Yet, I renamed this data-member to `pseudoKey`.

File: src/narrowdown.h (hunk 2 above)
Comment (veloman-yunkan) [reply src/narrowdown.h]: I had a rough idea of how this is going to work when it's ready, however I had to adjust it while understanding better all the subtleties of that approach. The latest state of the code ended up with a different semantic and name for this field.

**Comment by veloman-yunkan** [comment]
@mgautierfr The updated description contains some useful information (I think that it makes sense to copy some of it into the code).

**Comment by kelson42** [comment]
@mgautierfr @veloman-yunkan This is important to move on this because of the overall project(s) timeline but as well because discussing too long has not really proven to be that much beneficial in the past. So, at least for that PR, please try to be focused on the essential and would be great if we don't have days of latencies beetween each exchange. Would be really appreciated if this could be merged this week.

File: src/narrowdown.h [hunk 4]
```diff
@@ -81,9 +81,9 @@ class NarrowDown
 
   static std::string shortestStringInBetween(const std::string& a, const std::string& b)
   {
-    ASSERT(a, <, b);
+    ASSERT(a, <=, b);
     const auto m = std::mismatch(a.begin(), a.end(), b.begin());
-    return std::string(b.begin(), m.second+1);
+    return std::string(b.begin(), std::min(b.end(), m.second+1));
```
Comment (mgautierfr) [inline src/narrowdown.h:90]: This comment is about the whole commit, not the code but I must comment on code.

I'm not sure we need this (or that we should handle it this way).
URl's items must be unique. If not, this is a error in the zim file.
Older implementations of libzim creator were not enforce this but "newer" one does.
So no zimfile should come in this situation. Only old ones may and we should consider them as broken.

I didn't know about a zim file with this problem. Now we have one and it is a "fake" one.
We should probably raise a ZimFormatError here.

**Review by mgautierfr — COMMENTED** [review]
This PR is pretty good !!

We may move the PR's description in the commits (commit message at least, maybe code) but there is nothing to say on the code.

This case algorithm uses far less memory than my proposal of the fanout algorithm. We may have to discuss about fanout later (if we need it) but for now this PR is ok.

Here some open questions (the first one to be answered before merging) :
- The commit about not unique url (see my bellow comment on code)
- Should we still use a dirent cache ? We have greatly improve  the cache performance with previous commits but it seems that this cache may become useless as we are greatly reducing the search range and there are less chances that we search in the same range again.
- We may introduce a cache `url`->`dirent` instead of the current one (`index`->`dirent`) as some content may be load several times (main pages, css, ...)
- Should we keep the namespace cache ? Now we use it only to count the number of articles in each namespace. We don't use the boundary to search content.

File: src/narrowdown.h (hunk 4 above)
Comment (veloman-yunkan) [reply src/narrowdown.h:142]: I don't have any preference with regard to that detail. Please feel free to change the code as you like.

**Comment by veloman-yunkan** [comment]
> * Should we still use a dirent cache ? We have greatly improve  the cache performance with previous commits but it seems that this cache may become useless as we are greatly reducing the search range and there are less chances that we search in the same range again.
> 
> * We may introduce a cache `url`->`dirent` instead of the current one (`index`->`dirent`) as some content may be load several times (main pages, css, ...)
> 
> * Should we keep the namespace cache ? Now we use it only to count the number of articles in each namespace. We don't use the boundary to search content.

Can we please discuss those in a separate ticket?

File: src/dirent_lookup.h (hunk 3 above)
Comment (veloman-yunkan) [reply src/dirent_lookup.h:170]: I had the same idea for compressing the key storage even further by encoding the shared prefixes in a special way, which would also automatically eliminate the shared prefix from the string comparison. But I decided not to include that optimization in this PR.

**Comment by veloman-yunkan** [comment]
@kelson42 @mgautierfr Do you have any idea how the failure of the Macos CI builds can be fixed? For all of my recent pushes in this PR the Macos builds failed during the **Install packages** step (for example, [Macos (native_dyn)](https://github.com/openzim/libzim/pull/430/checks?check_run_id=1314738264)):

```
Run mstksg/get-package@v1
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Error: Fetching /usr/local/Homebrew/Library/Taps/local/homebrew-openssl failed!
Fetching /usr/local/Homebrew/Library/Taps/local/homebrew-python2 failed!
fatal: invalid upstream 'origin/master'
fatal: invalid upstream 'origin/master'
Error: Command failed: brew update && brew install gcovr pkg-config ninja
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Error: Fetching /usr/local/Homebrew/Library/Taps/local/homebrew-openssl failed!
Fetching /usr/local/Homebrew/Library/Taps/local/homebrew-python2 failed!
fatal: invalid upstream 'origin/master'
fatal: invalid upstream 'origin/master'
```

**Comment by kelson42** [comment]
@veloman-yunkan Not really, I have fixed one source of the problem (in git master already since a few days)... but now there is an other sympton.

File: src/narrowdown.h (hunk 4 above)
Comment (mgautierfr) [reply src/narrowdown.h:142]: Let's keep it. Even if the zim file is not theoretically valid, it will allow the user to read most of the entries.

**Review by mgautierfr — APPROVED** [review]
I approve the changes.
Please `rebase -i` to squash the `fixup!` commits before merging and we are good.

**Comment by mgautierfr** [comment]
> @kelson42 @mgautierfr Do you have any idea how the failure of the Macos CI builds can be fixed? 

I'm on it.
It seems that the github actions we use to install packages is failing. Directly using brew command work.
I'm update the CI in all our repositories.

**Comment by veloman-yunkan** [comment]
> Please `rebase -i` to squash the `fixup!` commits before merging and we are good.

@mgautierfr Done

**Comment by kelson42** [comment]
@mgautierfr @veloman-yunkan I think we can merge now.

**Pull request merged**